### PR TITLE
Dgraph Swarm Cluster in Pulumi

### PR DESCRIPTION
**Pull request by wimax-grapl**
### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/441

### What changes does this PR make to Grapl? Why?
Adds a new pulumi resource for a Dgraph swarm cluster.
Doesn't happen locally, only in AWS, since we're killing localgrapl.

### How were these changes tested?
The deploy to AWS works fine, but I haven't tested it against the provisioner lambda (that's not in pulumi yet, but it's up next)

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -470,7 +470,7 @@ services:
       /bin/bash -c "
          pulumi login --local &&
          pulumi stack init local-grapl --non-interactive &&
-         pulumi up --yes --skip-preview --stack=local-grapl &&
+         pulumi up --yes --skip-preview --stack=local-grapl --verbose 5 &&
```
Comment (wimax-grapl) [inline docker-compose.yml:473]: just for the sake of attempting to debug it. does nothing

File: pulumi/infra/swarm.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+                        ],
+                        "Resource": "*",
+                    },
+                ],
+            }
+        ),
+        opts=opts,
+    )
+    return aws.iam.RolePolicyAttachment(
+        f"attach_{policy_name}",
```
Comment (wimax-grapl) [inline pulumi/infra/swarm.py:102]: (this is me manually attempting to get around an un-needed `apply`; god i'd so much prefer some sort of await type syntax here)

File: pulumi/infra/swarm.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        # InstanceProfile for swarm instances
+        aws.iam.InstanceProfile(
+            "SwarmInstanceProfile",
+            opts=child_opts,
+            role=self.role.name,
+            name=f"{DEPLOYMENT_NAME}-swarm-instance-profile",
+        )
+
+    def cluster_host_port(self) -> pulumi.Output[str]:
+        return Output.concat(self.swarm_hosted_zone.name, ":9080")
```
Comment (christophermaier) [inline pulumi/infra/swarm.py:170]: Would it be worth extracting this port as a variable? It's repeated 3 times in various places in this code.

File: pulumi/infra/swarm.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+                "_to_",
+                self.security_group._name,
+                "_for_",
+                str(port_range),
+            ]
+        )
+
+        # We'll accept connections from Other into SecurityGroup
+        aws.ec2.SecurityGroupRule(
+            f"ingress_{descriptor}",
```
Comment (christophermaier) [inline pulumi/infra/swarm.py:188]: Naming and casing.

File: pulumi/infra/swarm.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+            opts=opts,
```
Comment (christophermaier) [inline pulumi/infra/swarm.py:189]: I think it might make sense to have the SecurityGroup be the parent resource:

```python
...
opts=pulumi.ResourceOptions(parent=self.security_group)
...
```
Incidentally, that might allow you to simplify the name of the `SecurityGroupRule`; you wouldn't necessarily need the name of `self.security_group` because that could then be understood from the parent/child relationship.

(Same applies for the egress rules below.)

File: pulumi/infra/swarm.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+        self, other: aws.ec2.SecurityGroup, port_range: Ec2Port, opts: ResourceOptions
+    ) -> None:
+        descriptor = "".join(
+            [
+                "from_",
+                other._name,
+                "_to_",
+                self.security_group._name,
+                "_for_",
+                str(port_range),
```
Comment (christophermaier) [inline pulumi/infra/swarm.py:182]: Might want a custom string method for this; as it is, this generates a name with content like `Ec2Port(protocol='tcp', port=9080)`; something like `TCP:9080` would be just as informative, but less verbose.

File: pulumi/infra/swarm.py [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+            to_port=port_range.port,
+            protocol=port_range.protocol,
+        )
+
+        # Allow connections out of Other to Self
+        aws.ec2.SecurityGroupRule(
+            f"egress_{descriptor}",
+            opts=opts,
+            type="egress",
+            # Perhaps this is wrong
```
Comment (christophermaier) [inline pulumi/infra/swarm.py:203]: Why?

File: pulumi/__main__.py [hunk 8]
```diff
@@ -3,22 +3,35 @@
...
 from infra.service_queue import ServiceQueue
 
-if __name__ == "__main__":
 
+def _get_dgraph_cluster(network: Network) -> DgraphCluster:
+    if LOCAL_GRAPL:
+        return LocalStandInDgraphCluster()
+    else:
+        return DgraphCluster(
+            name="swarm",
```
Comment (christophermaier) [inline pulumi/__main__.py:20]: "dgraph" might be a better name for this than "swarm", since the latter is an implementation detail.

File: pulumi/__main__.py [hunk 9]
```diff
@@ -3,22 +3,35 @@
...
 from infra.dgraph_ttl import DGraphTTL
 from infra.engagement_creator import EngagementCreator
 from infra.metric_forwarder import MetricForwarder
 from infra.network import Network
 from infra.secret import JWTSecret
 from infra.service_queue import ServiceQueue
 
-if __name__ == "__main__":
 
+def _get_dgraph_cluster(network: Network) -> DgraphCluster:
```
Comment (christophermaier) [inline pulumi/__main__.py:15]: For overall clarity, I'd recommend calling this something like `_create_dgraph_cluster`. In Pulumi, getter functions are pretty exclusively associated with retrieving a reference to already existing infrastructure, which this is definitely not doing.

File: pulumi/__main__.py [hunk 10]
```diff
@@ -89,3 +107,7 @@
         user.local_grapl_user(
             dynamodb_tables.user_auth_table, "grapluser", "graplpassword"
         )
+
+
+if __name__ == "__main__":
+    main()
```
Comment (christophermaier) [inline pulumi/__main__.py:113]: Nice 👍

File: pulumi/infra/bucket.py [hunk 11]
```diff
@@ -28,13 +30,14 @@ def __init__(
         opts: `pulumi.ResourceOptions` for this resource.
 
         """
-        physical_bucket_name = bucket_physical_name(logical_bucket_name)
+        self.logical_bucket_name = logical_bucket_name
```
Comment (christophermaier) [inline pulumi/infra/bucket.py:33]: Why was this needed?

File: pulumi/infra/connectable.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+from typing import Optional, Sequence
+
+import pulumi_aws as aws
+from pulumi_aws.ec2 import outputs
+from typing_extensions import Protocol
+
+import pulumi
+
+
+class SwarmConnectable(Protocol):
```
Comment (christophermaier) [inline pulumi/infra/connectable.py:12]: Is this used anywhere?

File: pulumi/infra/policies.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+                            "Effect": "Allow",
+                            "Action": [
+                                "logs:CreateLogGroup",
+                                "logs:CreateLogStream",
+                                "logs:PutLogEvents",
+                                "logs:DescribeLogStreams",
+                            ],
+                            "Resource": f"{arn}:*",
+                        }
+                    ],
```
Comment (christophermaier) [inline pulumi/infra/policies.py:145]: Looking closer, this policy appears to be a strict subset of the managed policy created above, and we attach it to the same Role. I don't think this is actually needed.

File: pulumi/infra/swarm.py [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+    )
+
+
+class Swarm(pulumi.ComponentResource):
+    def __init__(
+        self,
+        name: str,
+        vpc: Vpc,
+        internal_service_ports: Sequence[Ec2Port],
+        log_group: aws.cloudwatch.LogGroup,
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply pulumi/infra/swarm.py]: blindly cloning existing infra instead of thinking about it :)

File: pulumi/infra/swarm.py [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+            vpcs=[
+                aws.route53.ZoneVpcArgs(
+                    vpc_id=vpc.id,
+                )
+            ],
+            opts=child_opts,
+        )
+
+        self.swarm_config_bucket = Bucket(
+            logical_bucket_name="swarm-config-bucket",
```
[the remark this answers is not in the document]
Comment (wimax-grapl) [reply pulumi/infra/swarm.py:161]: this one has to be fixed, for now, for graplctl. May change in followup PR.

File: pulumi/infra/swarm.py (hunk 3 above)
Comment (wimax-grapl) [reply pulumi/infra/swarm.py:171]: Ehhhhh it's repeated 3 times because of throwaway localgrapl stuff

File: pulumi/infra/swarm.py (hunk 6 above)
Comment (wimax-grapl) [reply pulumi/infra/swarm.py:183]: yeah, sure

File: pulumi/infra/swarm.py (hunk 5 above)
Comment (wimax-grapl) [reply pulumi/infra/swarm.py]: i like it

File: pulumi/infra/swarm.py (hunk 7 above)
Comment (wimax-grapl) [reply pulumi/infra/swarm.py]: adding a comment

File: pulumi/infra/swarm.py (hunk 6 above)
Comment (jgrillo-grapl) [reply pulumi/infra/swarm.py:183]: probably both `__str__` and `__repr__` should be like `f"{protocol}:{port}"`?

**Review by jgrillo-grapl — APPROVED** [review]
LGTM

File: pulumi/infra/swarm.py (hunk 6 above)
Comment (christophermaier) [reply pulumi/infra/swarm.py:183]: Yeah, that's more what I was thinking, too.

`egress-from-graphql-endpoint-security-group-for-tcp:9080`
is nicer than
`egress-from-graphql-endpoint-security-group-for-Ec2Port(tcp:9080)`

**Review by christophermaier — APPROVED** [review]
Nice stuff!

**Comment by wimax-grapl** [comment]
i feel like 9/10 times you don't need to update the repr,  shrug
